.gitlab-ci.yml

```diff
@@ -1,86 +1,175 @@
+  FORCE_DEPLOY:               ""        # boolean: true or false - triggers the deploy-production stage
+  FORCE_DOCKER_TAG:           ""        # choose an existing docker tag to be deployed (e.g. v1.2.3)
+
+  # Vault variables (Optional)
+  VAULT_SERVER_URL:           "https://vault.parity-mgmt-vault.parity.io"
+  VAULT_AUTH_PATH:            "gitlab-parity-io-jwt"
+  VAULT_AUTH_ROLE:            "cicd_gitlab_parity_${CI_PROJECT_NAME}"
+
+
+default:
+  before_script:
+    - |-
+      echo defining DOKCER_IMAGE_TAG variable
+      if [[ $FORCE_DOCKER_TAG ]]; then
+        export DOCKER_IMAGE_TAG="${FORCE_DOCKER_TAG}"
+      elif [[ $CI_COMMIT_TAG =~ ^v[0-9]+\.[0-9]+.*$ ]]; then
+        export DOCKER_IMAGE_TAG="${CI_COMMIT_TAG}"
+        #export BUILD_LATEST_IMAGE="true"
+      else
+        export DOCKER_IMAGE_TAG="${CI_COMMIT_SHORT_SHA}-beta"
+      fi
 
 stages:
-  - dockerize
-  - staging
+  - build
+  - deploy-staging
+  - deploy-production
+
 
-.vault-secrets:                    &vault-secrets
+# Pipeline Job Templates:
+.vault-secrets:           &vault-secrets
   secrets:
     DOCKER_HUB_USER:
-      vault:                       cicd/gitlab/parity/DOCKER_HUB_USER@kv
-      file:                        false
+      vault:              cicd/gitlab/parity/DOCKER_HUB_USER@kv
+      file:               false
     DOCKER_HUB_PASS:
-      vault:                       cicd/gitlab/parity/DOCKER_HUB_PASS@kv
-      file:                        false
+      vault:              cicd/gitlab/parity/DOCKER_HUB_PASS@kv
+      file:               false
 
 .dockerize:               &dockerize
-  stage:                  dockerize
+  stage:                  build
   image:                  quay.io/buildah/stable
+  script:
+    - |-
+      echo building "$CONTAINER_REPO:$DOCKER_IMAGE_TAG"
+      if [[ $BUILD_LATEST_IMAGE ]]; then
+        buildah bud \
+        --format=docker \
+        --tag "$CONTAINER_REPO:$DOCKER_IMAGE_TAG" \
+        --tag "$CONTAINER_REPO:latest" "$DOCKERFILE_DIRECTORY"
+      else
+        buildah bud \
+        --format=docker \
+        --tag "$CONTAINER_REPO:$DOCKER_IMAGE_TAG" "$DOCKERFILE_DIRECTORY"
+      fi
+    - echo ${DOCKER_HUB_PASS} |
+        buildah login --username ${DOCKER_HUB_USER} --password-stdin docker.io
+    - |-
+      echo pushing "$CONTAINER_REPO:$DOCKER_IMAGE_TAG"
+      if [[ $BUILD_LATEST_IMAGE ]]; then
+        buildah push --format=v2s2 "$CONTAINER_REPO:$DOCKER_IMAGE_TAG"
+        buildah push --format=v2s2 "$CONTAINER_REPO:latest"
+      else
+        buildah push --format=v2s2 "$CONTAINER_REPO:$DOCKER_IMAGE_TAG"
+      fi
   rules:
-    - if: '$CI_COMMIT_BRANCH == "master"'
+    - if: '$FORCE_DOCKER_TAG'
+      when: never
+    - if: '$CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+.*$/'         # i.e. v1.0, v2.1rc1
+      when: always
+    - if: '$CI_COMMIT_REF_NAME == "master" || $CI_COMMIT_REF_NAME == "main"'
       when: always
```

Comment:
do the reuls get evaluated one after the other, so if FORCE_DOCKER_TAG is true, we'll never build the images, or do they all get individually checked and _any_ true result leads to the step being run?

Reply:
>  so if FORCE_DOCKER_TAG is true, we'll never build the images
yes, exactly!

Reply:
Cool, thanks!